utils/terrain/src/map.rs

```diff
@@ -0,0 +1,217 @@
+use std::mem;
+use std::ops::{Index, IndexMut};
+
+/// A simple "flat" map based on a sorted vector
+///
+/// See the [module level documentation][super] for why one should use this.
+///
+/// The API is roughly similar to that of [`std::collections::HashMap`], though it
+/// requires `Ord` instead of `Hash`.
+#[derive(Clone, Debug)]
+pub struct VecMap<K, V> {
+    values: Vec<(K, V)>
```

Comment:
What @echeran suggested could be a future optimization with replacing the `Vec` here with https://crates.io/crates/smallvec . Definitely a follow up :)